packages/@sanity/types/src/schema/types.ts

```diff
@@ -717,28 +186,20 @@ export interface TextSchemaType extends StringSchemaType {
 
 export interface NumberSchemaType extends BaseSchemaType {
   jsonType: 'number'
-  options?: EnumListProps<number>
+  options?: NumberOptions
   initialValue?: InitialValueProperty<any, number>
 }
 
 export interface BooleanSchemaType extends BaseSchemaType {
   jsonType: 'boolean'
-  options?: {
-    layout: 'checkbox' | 'switch'
-  }
+  options?: BooleanOptions
   initialValue?: InitialValueProperty<any, boolean>
 }
 
 export interface ArraySchemaType<V = unknown> extends BaseSchemaType {
   jsonType: 'array'
   of: (Exclude<SchemaType, ArraySchemaType> | ReferenceSchemaType)[]
-  options?: {
-    list?: TitledListValue<V>[] | V[]
-    layout?: V extends string ? 'tags' : 'grid'
-    direction?: 'horizontal' | 'vertical'
-    sortable?: boolean
-    modal?: {type?: 'dialog' | 'popover'; width?: number | 'auto'}
-  }
+  options?: ArrayOptions<V> & {layout?: V extends string ? 'tag' : 'grid'}
```

Comment:
Had to diverge from the definition type here, and narrow layout a bit. This is more correct, but supporting it in the definition is not worth it (need to drill the type of array.of every where via generics)